### PR TITLE
CDAP-14022 add more bootstrap step types

### PR DESCRIPTION
**Pull request by albertshau**
Add implementations for the rest of the bootstrap step types for
creating applications, starting programs, and setting system
preferences.

Updated the bootstrap files to create the dataprep application
and to start the service in sandbox. Removed the internal
services that were responsible for creating and starting dataprep
in sandbox.

Also added the bootstrap rest api, which simply reloads the
config and runs all the steps.

### CODE REVIEW

**Comment by albertshau** [comment]
https://builds.cask.co/browse/CDAP-DUT6626

File: cdap-api/src/main/java/co/cask/cdap/api/artifact/ArtifactSummary.java [hunk 1]
```diff
@@ -67,6 +67,24 @@ public ArtifactScope getScope() {
...
+   * user input.
+   *
+   * @throws IllegalArgumentException if the artifact summary is invalid
+   */
+  public void validate() {
+    if (name == null) {
+      throw new IllegalArgumentException("Artifact name must be specified.");
+    }
+    if (version == null) {
+      throw new IllegalArgumentException("Artifact version must be specified.");
```
Comment (gsps1) [inline cdap-api/src/main/java/co/cask/cdap/api/artifact/ArtifactSummary.java]: might be better to say `Artifact version is missing for arfitfact {}, artifactNAme`

File: cdap-app-fabric/src/main/java/co/cask/cdap/config/PreferencesService.java [hunk 2]
```diff
@@ -90,60 +94,84 @@ public PreferencesService(DatasetFramework datasetFramework, TransactionSystemCl
...
-      // After everything is set, publish the update message and add the association if profile is present
-      if (profile.isPresent()) {
-        profileDataset.addProfileAssignment(profile.get(), entityId);
-        adminEventPublisher.publishProfileAssignment(entityId, profile.get());
+      if (cause instanceof ProfileConflictException) {
+        throw (ProfileConflictException) cause;
       }
+      if (cause instanceof BadRequestException) {
+        throw (BadRequestException) cause;
+      }
```
Comment (gsps1) [inline cdap-app-fabric/src/main/java/co/cask/cdap/config/PreferencesService.java:116]: why the change in pattern from `transactionals.execute` that can accept exceptions to this ?

File: cdap-app-fabric/src/main/java/co/cask/cdap/config/PreferencesService.java [hunk 3]
```diff
@@ -90,60 +94,84 @@ public PreferencesService(DatasetFramework datasetFramework, TransactionSystemCl
...
+    NamespaceId namespaceId = isInstanceLevel ?
+      NamespaceId.SYSTEM : ((NamespacedEntityId) entityId).getNamespaceId();
+
+    // validate the profile and publish the necessary metadata change if the profile exists in the property
+    Optional<ProfileId> profile = SystemArguments.getProfileIdFromArgs(namespaceId, propertyMap);
+    if (profile.isPresent()) {
+      ProfileId profileId = profile.get();
+      // if it is instance level set, the profile has to be in SYSTEM scope, so throw BadRequestException when
+      // setting a USER scoped profile
+      if (isInstanceLevel && !propertyMap.get(SystemArguments.PROFILE_NAME).startsWith(EntityScope.SYSTEM.name())) {
```
Comment (gsps1) [inline cdap-app-fabric/src/main/java/co/cask/cdap/config/PreferencesService.java:137]: could this throw NPE when the profile name doesn't exist in propertyMap, or is it guaranteed to be present.

File: cdap-app-fabric/src/main/java/co/cask/cdap/config/PreferencesService.java [hunk 4]
```diff
@@ -90,60 +94,84 @@ public PreferencesService(DatasetFramework datasetFramework, TransactionSystemCl
...
+    }
+
+    // need to get old property and check if it contains profile information
+    Map<String, String> oldProperties = preferencesDataset.getPreferences(entityId);
+    // get the old profile information from the previous properties
+    Optional<ProfileId> oldProfile = SystemArguments.getProfileIdFromArgs(namespaceId, oldProperties);
+    preferencesDataset.setPreferences(entityId, propertyMap);
+
+    // After everything is set, publish the update message and add the association if profile is present
+    if (profile.isPresent()) {
```
Comment (gsps1) [inline cdap-app-fabric/src/main/java/co/cask/cdap/config/PreferencesService.java:160]: probably just due to refactoring, but this block can be in the one above `isPresent` where we throw exceptions  right ?

File: cdap-app-fabric/src/main/java/co/cask/cdap/config/PreferencesService.java [hunk 5]
```diff
@@ -220,31 +248,69 @@ private void deleteConfig(EntityId entityId) {
...
+
+    Set<String> added = new HashSet<>(properties.keySet());
+    try {
+      transactional.execute(context -> {
+        ProfileDataset profileDataset = ProfileDataset.get(context, datasetFramework);
+        PreferencesDataset preferencesDataset = PreferencesDataset.get(context, datasetFramework);
+        Map<String, String> oldProperties = preferencesDataset.getPreferences(instanceId);
+        Map<String, String> newProperties = new HashMap<>(properties);
+        newProperties.putAll(oldProperties);
+        added.removeAll(oldProperties.keySet());
```
Comment (gsps1) [inline cdap-app-fabric/src/main/java/co/cask/cdap/config/PreferencesService.java]: might be better to read if we use exists check to add than to add and remove from set.
```
        properties.entrySet().stream().forEach(k -> {
          if (!oldProperties.containsKey(k.getKey())) {
            oldProperties.put(k.getKey(), k.getValue());
            added.add(k.getKey());
          }
        });
```

File: cdap-app-fabric/src/main/java/co/cask/cdap/gateway/handlers/BootstrapHttpHandler.java [hunk 6]
```diff
@@ -0,0 +1,50 @@
...
+  private final BootstrapService bootstrapService;
+
+  @Inject
+  BootstrapHttpHandler(BootstrapService bootstrapService) {
+    this.bootstrapService = bootstrapService;
+  }
+
+  @POST
+  @Path("/bootstrap")
+  public void get(HttpRequest request, HttpResponder responder) throws InterruptedException {
```
Comment (gsps1) [inline cdap-app-fabric/src/main/java/co/cask/cdap/gateway/handlers/BootstrapHttpHandler.java]: rename method name to be `bootstrap` than get, misleading with the Http method being POST.

File: cdap-app-fabric/src/main/java/co/cask/cdap/internal/bootstrap/executor/AppCreator.java [hunk 7]
```diff
@@ -0,0 +1,127 @@
...
+      arguments.getOwnerPrincipal() == null ? null : new KerberosPrincipalId(arguments.getOwnerPrincipal());
+
+    // if we don't null check, it gets serialized to "null"
+    String configString = arguments.getConfig() == null ? null : GSON.toJson(arguments.getConfig());
+
+    try {
+      appLifecycleService.deployApp(appId.getParent(), appId.getApplication(), appId.getVersion(),
+                                    artifactSummary, configString, x -> { },
+                                    ownerPrincipalId, arguments.canUpdateSchedules());
+    } catch (NotFoundException | UnauthorizedException | InvalidArtifactException e) {
```
Comment (gsps1) [inline cdap-app-fabric/src/main/java/co/cask/cdap/internal/bootstrap/executor/AppCreator.java:73]: `ArtifactNotFoundException` should also be here

File: cdap-app-fabric/src/main/java/co/cask/cdap/internal/bootstrap/executor/ProgramStarter.java [hunk 8]
```diff
@@ -0,0 +1,120 @@
...
+    ProgramId programId = arguments.getId();
+
+    try {
+      programLifecycleService.getProgramSpecification(programId);
+    } catch (NotFoundException e) {
+      throw new IllegalArgumentException(String.format("Cannot start %s because it does not exist.", programId), e);
+    }
+
+    // do nothing if the program is already running
+    if (programLifecycleService.getProgramStatus(programId) != ProgramStatus.STOPPED) {
```
Comment (gsps1) [inline cdap-app-fabric/src/main/java/co/cask/cdap/internal/bootstrap/executor/ProgramStarter.java:54]: should we allow suspended too ?

File: cdap-app-fabric/src/main/java/co/cask/cdap/internal/bootstrap/executor/SystemPreferenceSetter.java [hunk 9]
```diff
@@ -0,0 +1,83 @@
...
+import com.google.common.annotations.VisibleForTesting;
+import com.google.inject.Inject;
+import org.apache.tools.ant.util.Retryable;
+
+import java.util.Collections;
+import java.util.HashMap;
+import java.util.Map;
+
+/**
+ * Creates a system profile if it doesn't already exist.
```
Comment (gsps1) [inline cdap-app-fabric/src/main/java/co/cask/cdap/internal/bootstrap/executor/SystemPreferenceSetter.java]: incorrect javadoc

**Review by gsps1 — COMMENTED** [review]
few comments/questions.

File: cdap-app-fabric/src/main/java/co/cask/cdap/config/PreferencesService.java (hunk 2 above)
Comment (albertshau) [reply cdap-app-fabric/src/main/java/co/cask/cdap/config/PreferencesService.java:116]: that method can only take two types of exceptions, so the BadRequestException was getting wrapped by a RuntimeException

File: cdap-app-fabric/src/main/java/co/cask/cdap/internal/bootstrap/executor/AppCreator.java (hunk 7 above)
Comment (albertshau) [reply cdap-app-fabric/src/main/java/co/cask/cdap/internal/bootstrap/executor/AppCreator.java:73]: its a subclass of NotFoundException

File: cdap-app-fabric/src/main/java/co/cask/cdap/internal/bootstrap/executor/ProgramStarter.java (hunk 8 above)
Comment (albertshau) [reply cdap-app-fabric/src/main/java/co/cask/cdap/internal/bootstrap/executor/ProgramStarter.java:54]: ProgramStatus is different than ProgramRunStatus and only has starting, running, and stopped. suspended counts as running I believe.

File: cdap-api/src/main/java/co/cask/cdap/api/artifact/ArtifactSummary.java (hunk 1 above)
Comment (albertshau) [reply cdap-api/src/main/java/co/cask/cdap/api/artifact/ArtifactSummary.java]: done

File: cdap-app-fabric/src/main/java/co/cask/cdap/config/PreferencesService.java (hunk 3 above)
Comment (albertshau) [reply cdap-app-fabric/src/main/java/co/cask/cdap/config/PreferencesService.java:137]: it's guaranteed to be present because profileId exists

File: cdap-app-fabric/src/main/java/co/cask/cdap/config/PreferencesService.java (hunk 4 above)
Comment (albertshau) [reply cdap-app-fabric/src/main/java/co/cask/cdap/config/PreferencesService.java:160]: good point, done

File: cdap-app-fabric/src/main/java/co/cask/cdap/config/PreferencesService.java (hunk 5 above)
Comment (albertshau) [reply cdap-app-fabric/src/main/java/co/cask/cdap/config/PreferencesService.java]: I'll make it clearer with Sets.difference().

File: cdap-app-fabric/src/main/java/co/cask/cdap/internal/bootstrap/executor/SystemPreferenceSetter.java (hunk 9 above)
Comment (albertshau) [reply cdap-app-fabric/src/main/java/co/cask/cdap/internal/bootstrap/executor/SystemPreferenceSetter.java]: done

File: cdap-app-fabric/src/main/java/co/cask/cdap/config/PreferencesService.java [hunk 10]
```diff
@@ -154,12 +159,6 @@ private void setConfig(ProfileDataset profileDataset, PreferencesDataset prefere
     Optional<ProfileId> oldProfile = SystemArguments.getProfileIdFromArgs(namespaceId, oldProperties);
     preferencesDataset.setPreferences(entityId, propertyMap);
```
Comment (gsps1) [inline cdap-app-fabric/src/main/java/co/cask/cdap/config/PreferencesService.java]: earlier these steps happened before we publish like mentioned in the comment, is it safe to publish before setting preferences ?

**Review by gsps1 — DISMISSED** [review]
just one question, lgtm otherwise

File: cdap-app-fabric/src/main/java/co/cask/cdap/config/PreferencesService.java (hunk 10 above)
Comment (albertshau) [reply cdap-app-fabric/src/main/java/co/cask/cdap/config/PreferencesService.java]: yeah, the message publish happens in a transaction so it's fine. The comment is confusing though I will change it

Comment (albertshau) [reply cdap-app-fabric/src/main/java/co/cask/cdap/config/PreferencesService.java]: actually, I think moving it up here made the flow more confusing, I'm going to move it back, even though it will result in an extra if condition.

File: cdap-app-fabric/src/main/java/co/cask/cdap/config/PreferencesService.java [hunk 11]
```diff
@@ -90,60 +95,85 @@ public PreferencesService(DatasetFramework datasetFramework, TransactionSystemCl
...
+        throw (BadRequestException) cause;
+      }
+    }
+  }
 
-      // if old properties has the profile, remove the association
-      if (oldProfile.isPresent()) {
-        profileDataset.removeProfileAssignment(oldProfile.get(), entityId);
+  /**
+   * Validate the profile status is enabled and set the preferences in same transaction
```
Comment (gsps1) [inline cdap-app-fabric/src/main/java/co/cask/cdap/config/PreferencesService.java]: update the javadoc as this method is not handling the transaction anymore.

File: cdap-app-fabric/src/main/java/co/cask/cdap/config/PreferencesService.java [hunk 12]
```diff
@@ -90,60 +95,85 @@ public PreferencesService(DatasetFramework datasetFramework, TransactionSystemCl
...
+        throw new ProfileConflictException(String.format("Profile %s in namespace %s is disabled. It cannot be " +
+                                                           "assigned to any programs or schedules",
+                                                         profileId.getProfile(), profileId.getNamespace()),
+                                           profileId);
       }
-    }, NotFoundException.class, ProfileConflictException.class);
+
+    }
+
+    // need to get old property and check if it contains profile information
```
Comment (gsps1) [inline cdap-app-fabric/src/main/java/co/cask/cdap/config/PreferencesService.java:153]: where is the check done ? is the check done when we check `oldProfile.isPresent` ?

**Review by gsps1 — COMMENTED** [review]
minor comments, but this method is hard to understand as the flow between comments and code is not natural.

File: cdap-app-fabric/src/main/java/co/cask/cdap/config/PreferencesService.java (hunk 12 above)
Comment (albertshau) [reply cdap-app-fabric/src/main/java/co/cask/cdap/config/PreferencesService.java:153]: yes

**Review by gsps1 — APPROVED** [review]
LGTM
